### PR TITLE
Adding support for multiple nuclei of the same order parameter

### PR DESCRIPTION
**Pull request by supriyoumich**
The changes are specific to nucleation model. As there can be numerous nuclei in a given simulation, it is computationally expensive to use separate order parameter for each of the nucleus. Therefore, multiple nuclei should be assigned to same order parameter which can reduce the computational cost, especially when the total number of nuclei is very large and simulation box is also large. But we need to make sure that the nuclei assigned to the same order parameter are far from each other in space and there is no possibility of coalescence during simulation. These changes are made in this commit. 

It will first make sure any two nuclei, irrespective of their order parameter, will not overlap (this was already there in the master code). Next, the code will check whether the distance between any two nuclei with same order parameter is less than a threshold distance. If the  condition is true, then nucleation is not allowed for these two nuclei. This will ensure that if same order parameter is used for multiple nuclei, then they are located far from each other at least by a threshold distance. The threshold distance can be changed via the input file.

### CODE REVIEW

**Comment by landinjm** [comment]
@supriyoumich Can you add a short description on what these commits do and why they're necessary?

**Comment by supriyoumich** [comment]
@landinjm description added.

File: include/userInputParameters.h [hunk 1]
```diff
@@ -250,6 +250,7 @@ class userInputParameters
   double                    buffer_between_grains;
 
   bool         load_grain_structure;
+  bool         load_unstructured_grid;
```
Comment (landinjm) [inline include/userInputParameters.h:253]: Why is this here?

File: include/userInputParameters.h [hunk 2]
```diff
@@ -232,11 +232,11 @@ class userInputParameters
...
-  bool                      evolution_before_nucleation;
-  // Declare later
-  // bool multiple_nuclei_per_order_parameter;
-  double min_distance_between_nuclei; // Only enforced for nuclei placed during
-                                      // the same time step
+
+  bool   multiple_nuclei_per_order_parameter;
+  double min_distance_between_nuclei; // Only enforced for nuclei placed during the same
+                                      // time step
+  double       min_distance_between_OP;
```
Comment (landinjm) [inline include/userInputParameters.h:239]: Can just use `min_distance_between_order_parameters`

File: include/userInputParameters.h (hunk 1 above)
Comment (supriyoumich) [reply include/userInputParameters.h:253]: It will not do harm. It is there because I broken up my all changes to do PR for multiple nuclei with same OP. This will be needed when I do the PR for vtk file read. But you can remove if it is making a problem.

File: src/inputFileReader/inputFileReader.cc [hunk 3]
```diff
@@ -635,7 +635,7 @@ inputFileReader::declare_parameters(dealii::ParameterHandler    &parameter_handl
                                   dealii::Patterns::Double(),
                                   "The time at which nucleation starts.");
   parameter_handler.declare_entry("Nucleation end time",
-                                  "1.0e10",
+                                  "1.0",
```
[the remark this answers is not in the document]
Comment (supriyoumich) [reply src/inputFileReader/inputFileReader.cc:638]: Okay, I was thinking about site saturation which will end in the beginning. But it was not important at all and 1E10 can be kept.

File: src/matrixfree/nucleation.cc [hunk 4]
```diff
@@ -190,132 +191,152 @@ MatrixFreePDE<dim, degree>::getLocalNucleiList(std::vector<nucleus<dim>> &newnuc
...
               unsigned int variable_index = userInputs.nucleating_variable_indices.at(i);
 
-              // Compute random no. between 0 and 1 (new method)
-              rand_val = distr(gen);
-              // Nucleation probability
-              double Prob = getNucleationProbability(variable_values,
-                                                     element_volume,
-                                                     ele_center,
-                                                     variable_index);
+              // NEW SECTION: Loop through each existing nuclei to verify if any of them
```
[the remark this answers is not in the document]
Comment (supriyoumich) [reply src/matrixfree/nucleation.cc:194]: Yeah this was written by David most probably. The srx code was mirrored from David's srx branch.

File: include/userInputParameters.h [hunk 5]
```diff
@@ -232,11 +232,11 @@ class userInputParameters
   bool                      nucleation_occurs;
   std::vector<unsigned int> nucleating_variable_indices;
   std::vector<unsigned int> nucleation_need_value;
-  bool                      evolution_before_nucleation;
```
[the remark this answers is not in the document]
Comment (supriyoumich) [reply include/userInputParameters.h:235]: What was the issue? is it undefined somewhere? This is created by David I think. I need to see what is the purpose.

File: include/userInputParameters.h (hunk 2 above)
Comment (supriyoumich) [reply include/userInputParameters.h:239]: Makes sense.
